main_test.go

```diff
@@ -0,0 +1,194 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+)
+
+func testRun(t *testing.T, wasmFile string, flags ...string) ([]byte, int) {
+	var logs bytes.Buffer
+	output := io.MultiWriter(testLogger(t), &logs)
+	exitCode := run(append([]string{"go_js_wasm_exec", wasmFile, "-test.v"}, flags...), output, testFlagSet(t))
+	return logs.Bytes(), exitCode
+}
+
+func testFlagSet(t *testing.T) *flag.FlagSet {
+	return flag.NewFlagSet("wasmbrowsertest", flag.ContinueOnError)
+}
+
+type testWriter struct {
+	testingT *testing.T
+}
+
+func testLogger(t *testing.T) io.Writer {
+	return &testWriter{t}
+}
+
+func (w *testWriter) Write(b []byte) (int, error) {
+	w.testingT.Helper()
+	w.testingT.Log(string(b))
+	return len(b), nil
+}
+
+// writeFile creates a file at $baseDir/$path with the given contents, where 'path' is slash separated
+func writeFile(t *testing.T, baseDir, path, contents string) {
+	t.Helper()
+	path = filepath.FromSlash(path)
+	fullPath := filepath.Join(baseDir, path)
+	err := os.MkdirAll(filepath.Dir(fullPath), 0755)
+	if err != nil {
+		t.Fatal("Failed to create file's base directory:", err)
+	}
+	err = os.WriteFile(fullPath, []byte(contents), 0600)
+	if err != nil {
+		t.Fatal("Failed to create file:", err)
+	}
+}
+
+// buildTestWasm builds the given Go package's test binary and returns the output Wasm file
+func buildTestWasm(t *testing.T, path string) string {
+	t.Helper()
+	outputFile := filepath.Join(t.TempDir(), "out.wasm")
+	cmd := exec.Command("go", "test", "-c", "-o", outputFile, ".")
+	cmd.Dir = path
+	cmd.Env = append(os.Environ(),
+		"GOOS=js",
+		"GOARCH=wasm",
+	)
+	output, err := cmd.CombinedOutput()
+	if len(output) > 0 {
+		t.Log(string(output))
+	}
+	if err != nil {
+		t.Fatal("Failed to build Wasm binary:", err)
+	}
+	return outputFile
+}
+
+func TestRunPassing(t *testing.T) {
+	t.Parallel()
```

Comment:
I feel like all of these tests might benefit from a higher level refactor and use a table-driven test, where you have the file contents, and the expected exit code as inputs, and then just pass them on to a driver function. It might look more readable that way. WDYT?

Reply:
Sounds good to me. I did initially have them separated so they'd be tightly tied to an commit/issue number in the Git blame, but I'm not very picky. Table-driven is cleaner anyway!

Fixed in 49ed4bf